crates/nu-command/src/commands/all.rs

```diff
@@ -53,72 +50,56 @@ impl WholeStreamCommand for Command {
     let condition = {
-        if block.block.block.len() != 1 {
-            return Err(ShellError::labeled_error(
-                "Expected a condition",
-                "expected a condition",
-                tag,
-            ));
+        if all_args.predicate.block.block.len() != 1 {
+            return err;
         }
-        match block.block.block[0].pipelines.get(0) {
+        match all_args.predicate.block.block[0].pipelines.get(0) {
             Some(item) => match item.list.get(0) {
                 Some(ClassifiedCommand::Expr(expr)) => expr.clone(),
                 _ => {
-                    return Err(ShellError::labeled_error(
-                        "Expected a condition",
-                        "expected a condition",
-                        tag,
-                    ));
+                    return err;
                 }
             },
             None => {
-                return Err(ShellError::labeled_error(
-                    "Expected a condition",
-                    "expected a condition",
-                    tag,
-                ));
+                return err;
             }
         }
     };
 
-    let init = Ok(InputStream::one(
-        UntaggedValue::boolean(true).into_value(&tag),
-    ));
+    let condition = condition.clone();
+    let scope = args.scope.clone();
 
-    Ok(input
-        .fold(init, move |acc, row| {
-            let condition = condition.clone();
-            let ctx = ctx.clone();
-            ctx.scope.enter_scope();
-            ctx.scope.add_vars(&block.captured.entries);
-            ctx.scope.add_var("$it", row);
-
-            let condition = evaluate_baseline_expr(&condition, &*ctx);
-            ctx.scope.exit_scope();
-
-            let curr = acc?.drain_vec();
-            let curr = curr
-                .get(0)
-                .ok_or_else(|| ShellError::unexpected("No value to check with"))?;
-            let cond = curr.as_bool()?;
-
-            match condition {
-                Ok(condition) => match condition.as_bool() {
-                    Ok(b) => Ok(InputStream::one(
-                        UntaggedValue::boolean(cond && b).into_value(&curr.tag),
-                    )),
-                    Err(e) => Err(e),
-                },
-                Err(e) => Err(e),
-            }
-        })?
-        .to_action_stream())
+    // Variables in nu are immutable. Having the same variable accross invocations
+    // of evaluate_baseline_expr does not mutate the variables and those each
+    // invocations are independent of each other!
+    scope.enter_scope();
+    scope.add_vars(&all_args.predicate.captured.entries);
+    let result: Result<bool, ShellError> = args.input.try_all(|row| {
+        //$it gets overwritten each invocation
+        scope.add_var("$it", row);
+
+        let condition = evaluate_baseline_expr(&*condition, &ctx);
+
+        Ok(condition?.as_bool()?)
+    });
+    scope.exit_scope();
+
+    Ok(OutputStream::one(result?))
```

Comment:
Not so sure, whether the original version was eagerly evaluated or not (I think it is?), but I also don't think it matters, as all is mostly at the end of a pipeline anyways.